qiskit_machine_learning/datasets/ad_hoc.py

```diff
@@ -31,240 +31,618 @@ def ad_hoc_data(
+         U_{\Phi(\vec{x})} \, H^{\otimes n} \, |0^{\otimes n}\rangle
 
     where
 
     .. math::
-        U_{{\Phi} (\vec{x})} = \exp \left( i \sum_{S \subseteq [n] } \phi_S(\vec{x})
-        \prod_{i \in S} Z_i \right)
+       U_{\Phi(\vec{x})}
+       = \exp\Bigl(i \sum_{S \subseteq [n]} \phi_S(\vec{x}) \prod_{i \in S} Z_i\Bigr),
 
     and
 
     .. math::
-        \begin{cases}
-        \phi_{\{i, j\}} = (\pi - x_i)(\pi - x_j) \\
-        \phi_{\{i\}} = x_i
-        \end{cases}
+       \phi_{\{i,j\}} = (\pi - x_i)(\pi - x_j), \quad
+       \phi_{\{i\}} = x_i.
+
+    The choice of second-order terms :math:`Z_i Z_j` in the sum depends on the
+    ``entanglement`` argument (e.g., ``"linear"``, ``"circular"``, or
+    ``"full"``).
 
-    We then attribute labels to the vectors according to the rule
+    An observable is then defined as
 
     .. math::
-        m(\vec{x}) = \begin{cases}
-        1 & \langle \Phi(\vec{x}) | V^\dagger \prod_i Z_i V | \Phi(\vec{x}) \rangle > \Delta \\
-        -1 & \langle \Phi(\vec{x}) | V^\dagger \prod_i Z_i V | \Phi(\vec{x}) \rangle < -\Delta
-        \end{cases}
+       O = V^\dagger \bigl(\prod_i Z_i\bigr) V,
 
-    where :math:`\Delta` is the separation gap, and
-    :math:`V\in \mathrm{SU}(4)` is a random unitary.
+    where :math:`V` is a random unitary matrix. Depending on the
+    ``labelling_method``:
 
-    The current implementation only works with n = 2 or 3.
+    - If ``"expectation"``, the expectation value
+      :math:`\langle \Phi(\vec{x})| O |\Phi(\vec{x})\rangle` is compared to the
+      gap parameter :math:`\Delta` (from ``gap``) to assign :math:`\pm 1` labels.
+    - If ``"measurement"``, a simple measurement in the computational basis is
+      performed to assign labels.
 
-    **References:**
+    **References**
 
     [1] Havlíček V, Córcoles AD, Temme K, Harrow AW, Kandala A, Chow JM,
-    Gambetta JM. Supervised learning with quantum-enhanced feature
-    spaces. Nature. 2019 Mar;567(7747):209-12.
+    Gambetta JM. *Supervised learning with quantum-enhanced feature spaces*.
+    Nature. 2019 Mar;567(7747):209–212.
     `arXiv:1804.11326 <https://arxiv.org/abs/1804.11326>`_
 
-    Args:
-        training_size: the number of training samples.
-        test_size: the number of testing samples.
-        n: number of qubits (dimension of the feature space). Must be 2 or 3.
-        gap: separation gap (:math:`\Delta`).
-        plot_data: whether to plot the data. Requires matplotlib.
-        one_hot: if True, return the data in one-hot format.
-        include_sample_total: if True, return all points in the uniform
-            grid in addition to training and testing samples.
+    Parameters
+    ----------
+    training_size : int
+        Number of training samples **per class**.
+    test_size : int
+        Number of testing samples **per class**.
+    n : int
+        Number of qubits (dimension of the feature space).
+    gap : int, optional
```

Comment:
```suggestion
    gap : int
```
'optional' can be misleading here. It's normally used when you have the syntax `gap: int | None  = None` in the declaration, and then something like `if not gap: gap = 0` in the function scope. Here, I'd just say that it defaults to 0 if not specified. Remember that this is an API change from the previous version, because `gap` goes from being an arg to a kwarg, so make sure you explain it in the release note.

Reply:
Same for the 'optional' in the other kwargs.